lucille-core/src/main/java/com/kmwllc/lucille/core/fileHandler/FileHandler.java

```diff
@@ -0,0 +1,113 @@
+import com.typesafe.config.Config;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * each implementation of the FileHandler handles a specific file type
+ */
+
+public interface FileHandler {
+
+  /**
+   * collection of file types that are supported by the FileHandler interface.
+   * Note that if you add a new file type to list, you must also add the corresponding handler in getNewFileHandler
+   */
+  Set<String> SUPPORTED_FILE_TYPES = Set.of("json", "jsonl", "csv", "xml");
+
+  /**
+   * processes a file given the Path to it and returns an iterator of Documents. The Iterator should close all resources
+   * when completed (hasNext() is false) or when an exception is thrown.
+   */
+  Iterator<Document> processFile(Path path) throws FileHandlerException;
+
+  /**
+   * processes a file given the file contents and representation path string to it and returns an iterator of Documents.
+   * The Iterator should close all resources when completed (hasNext() is false) or when an exception is thrown.
+   * Path string is used for populating file path field of document and for logging/error/debugging purposes.
+   */
+  Iterator<Document> processFile(byte[] fileContent, String pathStr) throws FileHandlerException;
+
+  /**
+   * performs any necessary setup before processing a file.
+   */
+  void beforeProcessingFile(Path path) throws FileHandlerException;
+
+  /**
+   * performs any necessary setup after processing a file.
+   */
+  void afterProcessingFile(Path path) throws FileHandlerException;
+
+  /**
+   * performs any necessary setup if handler encountered any error processing a file.
+   */
+  void errorProcessingFile(Path path);
+
+  /**
+   * performs any necessary setup before processing a file.
+   */
+  void beforeProcessingFile(byte[] content) throws FileHandlerException;
+
+  /**
+   * performs any necessary setup after processing a file.
+   */
+  void afterProcessingFile(byte[] content) throws FileHandlerException;
+
+  /**
+   * performs any necessary setup if handler encountered any error processing a file.
+   */
+  void errorProcessingFile(byte[] content);
+
+  /**
+   * A helper function that processes a file and publishes the documents to a Publisher
+   * using Path to file.
+   */
+  void processFileAndPublish(Publisher publisher, Path path) throws FileHandlerException;
+
+  /**
+   * A helper function that processes a file and publishes the documents to a Publisher
+   * using byte[] content.
+   */
+  void processFileAndPublish(Publisher publisher, byte[] fileContent, String pathStr) throws FileHandlerException;
+
+  /**
+   * getNewFileHandler returns a new FileHandler based on the file extension and file options.
+   * Note that if you add support for a new file type, you must also add the corresponding handler here
+   * AND in SUPPORTED_FILE_TYPES
+   */
+  static FileHandler getNewFileTypeHandler(String fileExtension, Config fileOptions) {
```

Comment:
suggest renaming to create()

Calling it would then look like: FileHandler.create("json", myConfig)

Reply:
changed to create() in [5d6a758](https://github.com/kmwtechnology/lucille/pull/217/commits/5d6a758f342c3994719490d6a22bb82c110d3545)